src/Microsoft.Azure.WebJobs.Host/Bindings/BindingContext.cs

```diff
@@ -110,7 +110,7 @@ internal AmbientBindingContext AmbientContext
             {
                 if (_ambientContext == null)
                 {
-                    _ambientContext = new AmbientBindingContext(_functionContext, _bindingData);
+                    _ambientContext = new AmbientBindingContext(_functionContext, () => BindingData);
```

Comment:
This was the bug - while we were correctly handling lazy evaluation when accessing the BindingData property, this property was not doing so when creating the AmbientBindingContext.